packages/rpc/src/vatidation/TracerResultParser.ts

```diff
@@ -0,0 +1,756 @@
+ * @param userOp the userOperation that was used in this simulation
+ * @param tracerResults the tracer return value
+ * @param validationResult output from simulateValidation
+ * @param entryPoint the entryPoint that hosted the "simulatedValidation" traced call.
+ * @return list of contract addresses referenced by this UserOp
+ */
+export function tracerResultParser(
+    userOp: UserOperation,
+    tracerResults: BundlerTracerResult,
+    validationResult: ValidationResult,
+    entryPointAddress: Address
+): [string[], StorageMap] {
+    // todo: block access to no-code addresses (might need update to tracer)
+
+    // opcodes from [OP-011]
+    const bannedOpCodes = new Set([
+        "GASPRICE",
+        "GASLIMIT",
+        "DIFFICULTY",
+        "TIMESTAMP",
+        "BASEFEE",
+        "BLOCKHASH",
+        "NUMBER",
+        "SELFBALANCE",
+        "BALANCE",
+        "ORIGIN",
+        "GAS",
+        "CREATE",
+        "COINBASE",
+        "SELFDESTRUCT",
+        "RANDOM",
+        "PREVRANDAO",
+        "INVALID"
+    ])
+
+    // eslint-disable-next-line @typescript-eslint/no-base-to-string
+    if (Object.values(tracerResults.callsFromEntryPoint).length < 1) {
+        throw new Error(
+            "Unexpected traceCall result: no calls from entrypoint."
+        )
+    }
+    const callStack = parseCallStack(tracerResults)
+
+    // [OP-052], [OP-053]
+    const callInfoEntryPoint = callStack.find(
+        (call) =>
+            call.to === entryPointAddress &&
+            call.from !== entryPointAddress &&
+            call.method !== "0x" &&
+            call.method !== "depositTo"
+    )
+    // [OP-054]
+    if (callInfoEntryPoint) {
+        throw new RpcError(
+            `illegal call into EntryPoint during validation ${callInfoEntryPoint?.method}`,
+            ValidationErrors.OpcodeValidation
+        )
+    }
+
+    // [OP-061]
+    const illegalNonZeroValueCall = callStack.find(
+        (call) =>
+            call.to !== entryPointAddress &&
+            hexToBigInt(call.value ?? "0x0") !== 0n
+    )
+
+    if (illegalNonZeroValueCall) {
+        throw new RpcError(
+            "May not may CALL with value",
+            ValidationErrors.OpcodeValidation
+        )
+    }
+
+    const sender = userOp.sender.toLowerCase()
+    // stake info per "number" level (factory, sender, paymaster)
+    // we only use stake info if we notice a memory reference that require stake
+    const stakeInfoEntities = {
+        factory: validationResult.factoryInfo,
+        account: validationResult.senderInfo,
+        paymaster: validationResult.paymasterInfo
+    }
+
+    const entitySlots: { [addr: string]: Set<string> } = parseEntitySlots(
+        stakeInfoEntities,
+        tracerResults.keccak as Hex[]
+    )
+
+    for (const [entityTitle, entStakes] of Object.entries(stakeInfoEntities)) {
+        const entityAddr = (entStakes?.addr ?? "").toLowerCase()
```

Comment:
yeah can we also not do lowercasing here?

Reply:
Here it is needed because we get verify it from the call stack and then we will have to parse the whole call stack too.